provider/metrics/container_insights.go

```diff
@@ -0,0 +1,268 @@
+	containerInsights.updateCumulativeValues(logger, pod, systemStats, netStats, &podStats)
+	return &podStats, nil
+}
+
+func collectMetrics(logger log.Logger, pod *v1.Pod, system, net *aci.ContainerGroupMetricsResult) stats.PodStats {
+	stat := stats.PodStats{
+		StartTime: pod.CreationTimestamp,
+		PodRef: stats.PodReference{
+			Name:      pod.Name,
+			Namespace: pod.Namespace,
+			UID:       string(pod.UID),
+		},
+		CPU: &stats.CPUStats{
+			UsageNanoCores:       newUInt64Pointer(0),
+			UsageCoreNanoSeconds: newUInt64Pointer(0),
+			Time:                 metav1.NewTime(time.Now()),
+		},
+		Memory: &stats.MemoryStats{
+			UsageBytes:      newUInt64Pointer(0),
+			WorkingSetBytes: newUInt64Pointer(0),
+			Time:            metav1.NewTime(time.Now()),
+		},
+		Network: &stats.NetworkStats{
+			Time: metav1.NewTime(time.Now()),
+			InterfaceStats: stats.InterfaceStats{
+				Name:     "eth0",
+				RxBytes:  newUInt64Pointer(0),
+				RxErrors: newUInt64Pointer(0),
+				TxBytes:  newUInt64Pointer(0),
+				TxErrors: newUInt64Pointer(0),
+			},
+		},
+		Containers: make([]stats.ContainerStats, 0),
+	}
+
+	for _, c := range pod.Status.ContainerStatuses {
+		containerName := c.Name
+		logger.Debugf("star to populate stats for container '%s'", containerName)
+		cs := stats.ContainerStats{
+			Name:      containerName,
+			StartTime: stat.StartTime,
+			CPU: &stats.CPUStats{
+				Time:                 metav1.NewTime(time.Now()),
+				UsageNanoCores:       newUInt64Pointer(0),
+				UsageCoreNanoSeconds: newUInt64Pointer(0),
+			},
+			Memory: &stats.MemoryStats{
+				Time:            metav1.NewTime(time.Now()),
+				RSSBytes:        newUInt64Pointer(0),
+				WorkingSetBytes: newUInt64Pointer(0),
+			},
+		}
+		if cpuSeries, found := findTimeSeries(system, aci.MetricTypeCPUUsage, &containerName); found && len(cpuSeries) > 0 {
+			data := cpuSeries[len(cpuSeries)-1]
+			nanoCores := uint64(data.Average * 1000000)
+			cs.CPU.UsageNanoCores = &nanoCores
+			cs.CPU.Time = metav1.NewTime(data.Timestamp)
+
+			podCPUCore := *stat.CPU.UsageNanoCores
+			podCPUCore += nanoCores
+			stat.CPU.UsageNanoCores = &podCPUCore
+			stat.CPU.Time = metav1.NewTime(data.Timestamp)
+		}
+		if memorySeries, found := findTimeSeries(system, aci.MetricTypeMemoryUsage, &containerName); found && len(memorySeries) > 0 {
+			data := memorySeries[len(memorySeries)-1]
+			bytes := uint64(data.Average)
+			cs.Memory.UsageBytes = &bytes
+			cs.Memory.WorkingSetBytes = &bytes
+			cs.Memory.Time = metav1.NewTime(data.Timestamp)
+
+			podMem := *stat.Memory.UsageBytes
+			podMem += bytes
+			stat.Memory.UsageBytes = &podMem
+			stat.Memory.WorkingSetBytes = &podMem
+			stat.Memory.Time = metav1.NewTime(data.Timestamp)
+		}
+		stat.Containers = append(stat.Containers, cs)
+	}
+
+	return stat
+}
+
+func findTimeSeries(metricsResult *aci.ContainerGroupMetricsResult, metricsType aci.MetricType, containerName *string) ([]aci.TimeSeriesEntry, bool) {
```

Comment:
fixed

Reply:
Yes, you are right. The codes of invoking Container Insight is not clear. I have encapsulated Container Insights related codes into containerInsightsMetricsWrapper, and as you suggested, I cached the result to avoid iterate multiple times.